website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,558 @@
+import Image from "next/image";
+
+At firezone, we build secure remote access that scales, be it from your Android phone, MacOS computer or Linux server.
+At the core of these apps sits a connectivity library - aptly named `connlib` - that manages network connections and wireguard tunnels to secure your traffic.
+After several iterations, we’ve landed on a design for our connectivity library that we are extremely happy with.
+It gives us blazingly fast and exhaustive tests, deep customisation and overall high assurance that it does what we want it to do.
+
+`connlib` is built in Rust and the design we are talking about is known as sans-IO.
+Rust's premise of speed and memory-safety makes it a great choice for building network services.
+Most parts of our Rust stack aren't particularly surprising:
+We use the `tokio` runtime for asynchronous tasks, `tungstenite` for WebSockets, `boringtun` for the WireGuard implementation, `rustls` to encrypt traffic with the API, etc.
+Yet, once you go beneath the surface of the library, you will discover something that is perhaps unusual:
+There are almost no calls to `tokio::spawn`, all communication is multiplexed via a single UDP socket and the same APIs appear to repeat themselves across various layers: `handle_timeout`, `poll_transmit`, `handle_input`, ...
+
+These are the tell-tales of a sans-IO design.
+Instead of sending & receiving bytes via a socket in multiple places, our protocols are implemented as pure state machines.
+Even time is abstracted away:
+Every function that needs to know the current time receives an `Instant` parameter instead of calling `Instant::now` itself.
+This pattern isn't something that we invented!
+The Python world even has a dedicated [website](https://sans-io.readthedocs.io/) about it.
+In Rust, it is used by libraries such as:
+
+- [`quinn`](https://github.com/quinn-rs/quinn/tree/main/quinn-proto), an independent QUIC implementation.
+- [`quiche`](https://github.com/cloudflare/quiche/tree/master/quiche), cloudflare's QUIC implementation.
+- [`str0m`](https://github.com/algesten/str0m), a sans-IO WebRTC implementation.
+
+In this post, we'll go over what it means to build a sans-IO networking library, why we think it is a good idea and lastly, why Rust lends itself particularly well to this pattern.
+
+## Rust's async model & the "function colouring" debate
+
```

Comment:
Might be worth elaborating on this section more with examples. It serves as our problem statement.

Reply:
You mean more examples in addition to the two? Or which concept do you think needs an example?